### PR TITLE
Refactor SonarAnalysisContext - Simplify inheritance

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Part of #6540

Remove inheritance from `SonarAnalysisContext` and `SonarParametrizedAnalysisContext`

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs [hunk 1]
```diff
@@ -125,4 +108,12 @@ public bool HasMatchingScope(DiagnosticDescriptor descriptor)
         bool ContainsTag(string tag) =>
             descriptor.CustomTags.Contains(tag);
     }
+
+    private ImmutableHashSet<string> CreateUnchangedFilesHashSet(AnalyzerOptions options) =>
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs:112]: `options` parameter is not used.

```suggestion
    private ImmutableHashSet<string> CreateUnchangedFilesHashSet() =>
```

File: analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs [hunk 2]
```diff
@@ -93,11 +54,30 @@ protected SonarAnalysisContextBase(SonarAnalysisContext analysisContext, TContex
         Context = context;
     }
 
-    public override bool TryGetValue<TValue>(SourceText text, SourceTextValueProvider<TValue> valueProvider, out TValue value) =>
-        AnalysisContext.TryGetValue(text, valueProvider, out value);
+    /// <param name="tree">Tree to decide on. Can be null for Symbol-based and Compilation-based scenarios. And we want to analyze those too.</param>
+    /// <param name="generatedCodeRecognizer">When set, generated trees are analyzed only when langauge-specific 'analyzeGeneratedCode' configuration property is also set.</param>
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs:58]:
```suggestion
    /// <param name="generatedCodeRecognizer">When set, generated trees are analyzed only when language-specific 'analyzeGeneratedCode' configuration property is also set.</param>
```

File: analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs [hunk 3]
```diff
@@ -93,11 +54,30 @@ protected SonarAnalysisContextBase(SonarAnalysisContext analysisContext, TContex
...
+                ? cachedProjectConfigReader
+                : throw new InvalidOperationException($"File '{Path.GetFileName(sonarProjectConfig.Path)}' has been added as an AdditionalFile but could not be read and parsed.");
+        }
+        else
+        {
+            return ProjectConfigReader.Empty;
+        }
+    }
 
     public bool IsScannerRun() =>
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs:82]: This is now used only in this file and can be made private.

File: analyzers/tests/SonarAnalyzer.UnitTest/AnalysisContext/SonarAnalysisContextBaseTest.ShouldAnalyzeTree.cs [hunk 4]
```diff
@@ -146,18 +141,18 @@ public void ShouldAnalyzeTree_GeneratedFile_AnalyzeGenerated_AnalyzeAllFiles(str
     public void ShouldAnalyzeTree_CorrectSettingUsed_VB(string fileName, bool expectedCSharp)
     {
         var sonarLintXml = CreateSonarLintXml(false);
-        var options = CreateOptions(sonarLintXml);
         var (compilationCS, treeCS) = CreateDummyCompilation(AnalyzerLanguage.CSharp, fileName);
         var (compilationVB, treeVB) = CreateDummyCompilation(AnalyzerLanguage.VisualBasic, fileName);
-        var sut = CreateSut();
+        var sutCS = CreateSut(compilationCS, CreateOptions(sonarLintXml));
+        var sutVB = CreateSut(compilationVB, CreateOptions(sonarLintXml));
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/AnalysisContext/SonarAnalysisContextBaseTest.ShouldAnalyzeTree.cs:147]: After the changes both cs and vb are tested but the test was meant only for VB. You need to update the test name or better extract another test for C# if needed.

File: analyzers/tests/SonarAnalyzer.UnitTest/AnalysisContext/SonarAnalysisContextBaseTest.ShouldAnalyzeTree.cs [hunk 5]
```diff
@@ -381,10 +373,10 @@ private static (Compilation Compilation, SyntaxTree Tree) CreateDummyCompilation
         return (compilation, compilation.SyntaxTrees.Single(x => x.FilePath.Contains(treeFileName)));
     }
 
-    private static bool ShouldAnalyzeTree(AnalyzerOptions options)
+    private bool ShouldAnalyzeTree(AnalyzerOptions options)
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/AnalysisContext/SonarAnalysisContextBaseTest.ShouldAnalyzeTree.cs:376]:
```suggestion
    private static bool ShouldAnalyzeTree(AnalyzerOptions options)
```

File: analyzers/tests/SonarAnalyzer.UnitTest/AnalysisContext/SonarAnalysisContextBaseTest.ShouldAnalyzeTree.cs [hunk 6]
```diff
@@ -381,10 +373,10 @@ private static (Compilation Compilation, SyntaxTree Tree) CreateDummyCompilation
...
     {
         var (compilation, tree) = CreateDummyCompilation(AnalyzerLanguage.CSharp, OtherFileName);
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/AnalysisContext/SonarAnalysisContextBaseTest.ShouldAnalyzeTree.cs:378]:
```suggestion
        var (_, tree) = CreateDummyCompilation(AnalyzerLanguage.CSharp, OtherFileName);
```

File: analyzers/tests/SonarAnalyzer.UnitTest/AnalysisContext/SonarAnalysisContextBaseTest.cs [hunk 7]
```diff
@@ -101,11 +101,86 @@ public void HasMatchingScope_MultipleDiagnostics_WithSingleScope_Scanner(bool ex
...
+    }
+
+    [TestMethod]
+    public void ProjectConfiguration_UsesCachedValue()
+    {
+        var options = TestHelper.CreateOptions($@"ResourceTests\SonarProjectConfig\Path_Windows\SonarProjectConfig.xml");
+        var firstSut = CreateSut(options);
+        var secondSut = CreateSut(options);
+        var firstConfig = firstSut.ProjectConfiguration();
+        var secondConfig = firstSut.ProjectConfiguration();
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/AnalysisContext/SonarAnalysisContextBaseTest.cs:120]:
```suggestion
        var secondConfig = secondSut.ProjectConfiguration();
```

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs:82]: Or removed as unneeded indirection :)

File: analyzers/tests/SonarAnalyzer.UnitTest/AnalysisContext/SonarAnalysisContextBaseTest.cs (hunk 7 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/AnalysisContext/SonarAnalysisContextBaseTest.cs:120]: ouch :disappear:

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**